src/lib.rs

```diff
@@ -398,36 +409,94 @@ impl TsunamiBuilder {
         let log = &this.log;
         let expected_num: u32 = this.descriptors.values().map(|&(_, n)| n).sum();
 
-        let all_the_things = futures::future::lazy(move || {
-            info!(log, "spinning up tsunami");
-
-            // set up network firewall for machines
-            let mut group_name = String::from("tsunami_security_");
-            group_name.extend(rand::thread_rng().gen_ascii_chars().take(10));
-            trace!(log, "creating security group"; "name" => &group_name);
-            let mut req = rusoto_ec2::CreateSecurityGroupRequest::default();
-            req.group_name = group_name;
-            req.description = "temporary access group for tsunami VMs".to_string();
-            ec2.create_security_group(&req)
-                .then(|r| r.context("failed to create security group for new machines"))
-        }).and_then(move |res| {
-            let group_id = res.group_id
-                .expect("aws created security group with no group id");
-            trace!(log, "created security group"; "id" => &group_id);
-
-            let mut req = rusoto_ec2::AuthorizeSecurityGroupIngressRequest::default();
-            req.group_id = Some(group_id.clone());
-
-            // ssh access
-            req.ip_protocol = Some("tcp".to_string());
-            req.from_port = Some(22);
-            req.to_port = Some(22);
-            req.cidr_ip = Some("0.0.0.0/0".to_string());
-            trace!(log, "adding ssh access to security group");
-            ec2.authorize_security_group_ingress(&req)
-                .then(|r| r.context("failed to fill in security group for new machines"))
-                .map(move |_| (group_id, req))
-        })
+        let all_the_things = futures::future::lazy(move || Self::configure_sg(log, ec2))
+            .and_then(move |group_id| {
+                // construct keypair for ssh access
+                Self::configure_ssh(log, ec2).map(|(pk, key_name)| (pk, key_name, group_id))
+            })
+            .and_then(move |(pk, key_name, group_id)| {
+                if this.cluster {
+                    Either::A(Self::configure_placement_groups(log, ec2))
+                } else {
+                    Either::B(futures::future::ok(None))
+                }.map(move |placement| (placement, pk, group_id, key_name))
+            })
+            .and_then(move |(placement, pk, group_id, key_name)| {
+                Self::issue_spot_requests(
+                    log,
+                    ec2,
+                    &this.descriptors,
+                    this.max_duration,
+                    placement,
+                    key_name,
+                    group_id,
+                ).map(
+                    |SpotRequestResult {
+                         id_to_name,
+                         setup_fns,
+                         usernames,
+                         spot_req_ids,
+                     }| (id_to_name, setup_fns, usernames, spot_req_ids, pk),
```

Comment:
I'd probably just pass along the entire `SpotRequestResult` here, and then deconstruct in the next future.